File: src/main/java/me/melijn/jda/blub/CommandClientBuilder.java
package me.melijn.jda.blub;

import me.melijn.jda.Melijn;

import java.util.HashSet;
import java.util.Set;

public class CommandClientBuilder {

    private final long ownerId;
    private final Set<Command> commands = new HashSet<>();
    private final Melijn melijn;

    public CommandClientBuilder(Melijn melijn, long ownerId) {
        this.melijn = melijn;
        this.ownerId = ownerId;
    }

    public CommandClient build() {
        return new CommandClientImpl(melijn, ownerId, commands);
    }

    public CommandClientBuilder addCommand(Command command) {
        commands.add(command);
        melijn.getMySQL().addCommand(command);
        return this;
    }
}

File: src/main/kotlin/me/melijn/melijnbot/objects/command/CommandClientBuilder.kt
package me.melijn.melijnbot.objects.command

import me.melijn.melijnbot.Container
import org.reflections.Reflections
import java.util.*


class CommandClientBuilder(private val container: Container) {

    private val commands = HashSet<AbstractCommand>()

    init {
        container.daoManager.commandWrapper.clearCommands()
    }

    fun build(): CommandClient {
        return CommandClient(commands.toSet(), container)
    }

    private suspend fun addCommand(command: AbstractCommand): CommandClientBuilder {
        commands.add(command)
        container.daoManager.commandWrapper.insert(command)
        return this
    }

    suspend fun loadCommands(): CommandClientBuilder {
        val reflections = Reflections("me.melijn.melijnbot.commands")

        val commands = reflections.getSubTypesOf(AbstractCommand::class.java)
        commands.forEach { command ->
            try {
                if (!command.isMemberClass) {
                    val cmd = command.getDeclaredConstructor().newInstance()
                    addCommand(cmd)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return this
    }
}